tests/registration_tests.c

```diff
@@ -17,12 +17,75 @@
  *******************************************************************************/
 
 #include "CUnit/CUnit.h"
+#include "connection.h"
+#include "er-coap-13/er-coap-13.h"
 #include "internals.h"
 #include "liblwm2m.h"
 #include "tests.h"
 
+#include <string.h>
+
 #ifdef LWM2M_SERVER_MODE
 
+static void create_test_registration_message(coap_packet_t *packet) {
+    coap_init_message(packet, COAP_TYPE_CON, COAP_POST, 0xbeef);
+    uint8_t token[] = {'t', 'o', 'k', 'e', 'n'};
+    coap_set_header_token(packet, token, sizeof(token));
+    coap_set_header_uri_path(packet, "rd");
+    coap_set_header_content_type(packet, LWM2M_CONTENT_LINK);
+    coap_set_header_uri_query(packet, "b=U&lwm2m=1.1&lt=300&ep=dummy-client");
+
+    char const *payload = "</1/0>,</3/0>;ver=\"1.2\"";
+    coap_set_payload(packet, payload, strlen(payload));
+}
+
+static void register_at_server(void) {
+    /* arrange */
+    coap_packet_t packet;
+    memset(&packet, 0, sizeof(packet));
+    create_test_registration_message(&packet);
+
+    uint8_t reg_coap_msg[80];
+    memset(reg_coap_msg, 0, sizeof(reg_coap_msg));
+    size_t msg_size = coap_serialize_message(&packet, reg_coap_msg);
+    CU_ASSERT_EQUAL(msg_size, 76);
+    const uint8_t expected_buffer[] = {
+        0x45, 0x02, 0xBE, 0xEF, 0x74, 0x6F, 0x6B, 0x65, 0x6E, 0xB2, 0x72, 0x64, 0x11, 0x28, 0x33, 0x62,
+        0x3D, 0x55, 0x09, 0x6C, 0x77, 0x6D, 0x32, 0x6D, 0x3D, 0x31, 0x2E, 0x31, 0x06, 0x6C, 0x74, 0x3D,
+        0x33, 0x30, 0x30, 0x0D, 0x02, 0x65, 0x70, 0x3D, 0x64, 0x75, 0x6D, 0x6D, 0x79, 0x2D, 0x63, 0x6C,
+        0x69, 0x65, 0x6E, 0x74, 0xFF, 0x3C, 0x2F, 0x31, 0x2F, 0x30, 0x3E, 0x2C, 0x3C, 0x2F, 0x33, 0x2F,
+        0x30, 0x3E, 0x3B, 0x76, 0x65, 0x72, 0x3D, 0x22, 0x31, 0x2E, 0x32, 0x22, 0x00, 0x00, 0x00, 0x00,
+    };
+    CU_ASSERT_NSTRING_EQUAL_FATAL(expected_buffer, reg_coap_msg, msg_size);
+    coap_free_header(&packet);
+
+    /* act */
+    lwm2m_context_t *const server_ctx = lwm2m_init(NULL);
+    lwm2m_handle_packet(server_ctx, reg_coap_msg, msg_size, NULL);
+    lwm2m_close(server_ctx);
+
+    /* assert */
+    size_t send_buffer_len;
+    uint8_t *send_buffer = test_get_send_buffer(&send_buffer_len);
```

Comment:
If I understand this correctly this buffer then contains the registration response? At first this was not clear to me. Maybe change the variable name to stress that it is the registration response we are checking here. Send buffer is a bit misleading in my opinion.